rfcs/20200306-single-client-parameter-server.md

```diff
@@ -0,0 +1,517 @@
+    out of order and function inputs are shared between sync groups. 
+
+    We don't support the cases where `args` or `kwargs` are bound to a specific
+    sync group. We will consider supporting them in the future.
+
+    If there are barriers in `replica_fn`, it is users' responsibility to make
+    sure they won't cause deadlock.
+    """
+    pass
+  
+  def join(self):
+    """Wait until all scheduled functions are finished.
+
+    Raises an error if any of the functions fails to execute. In this case,
+    there is no guarantee that non-failing functions will complete.
+
+    When join() is being called, it is not allowed to call `schedule`.
+    """
+    pass
+
+  def done(self):
+    """Returns True if there is no pending functions to be executed."""
+    pass
+
+  def local_results(self, future_list):
+    """Get concrete values of the future list.
+    
+    Poisoned future objects will give `None`.
+    """
+    pass
+```
+
+
+#### Custom Training Loop
+
+To construct a custom training loop, users need to
+
+
+*   use `strategy.experimental_distribute_datasets_from_function` to create one dataset per worker. At this point, we recommend against using `strategy.experimental_distribute_dataset`.
+*   create models under `strategy.scope` so variables will be assigned to parameter servers.
+*   likewise, create a Keras metric object under `strategy.scope`. We expect the metric variables to be stored on parameter servers. Each worker, within their `replica_fn`, updates the metric states.
+*   use `strategy.schedule` to schedule the `replica_fn` on one remote worker. This `replica_fn` should take an iterator and perform forward and backward computation. This `strategy.schedule` will only schedule this `replica_fn` and returns one or several `Future` objects immediately.  
+*   use `strategy.local_results` to get concrete values of results returned by `strategy.schedule`. This may be a blocking call if the result is not yet ready. With any failure that cannot be handled will be ignored and as a result some of the results may be `None`.
+*   call `strategy.join` to wait until all scheduled functions are executed.
+
+```python
+# Connect to remote servers with a user-provided `ClusterResolver` object.
+strategy = ParameterServerStrategyV2(cluster_resolver)
+
+dataset_fn = # a function that returns a dataset
+
+# Clone the dataset on all workers, shuffled with different seeds.
+distributed_dataset = strategy.experimental_distribute_datasets_from_function(
+      dataset_fn)
+
+with strategy.scope():
+  # Create variables on parameter servers in a round-robin fashion.
+  model = create_model()
+  optimizer = tf.keras.optimizers.Adam()
+  accuracy = tf.keras.metrics.CategoricalAccuracy(name="train_accuracy")
+
+  @tf.function
+  def replica_fn(iterator):
+    x, y = next(iterator)
+    with tf.GradientTape() as tape:
+      predictions = model(x, table, training=True)
+      loss = compute_loss(y, predictions)
```

Comment:
How do you think this should work with mixed precision loss scaling?

We can either have the PS store the loss scale or have each worker have it's own loss scale. If we also want to send gradients to the PS in fp16, this will become more difficult, as then we have to send scaled gradients to the PS, so the workers would also have to send the loss scale over.

Reply:
@reedwm Interesting question there. Just curious, how does current MWMS work with the loss scale?

Reply:
Currently for multi-worker mirrored strategies, each worker has its own independent loss scale. Each worker independently scales the loss, unscales the gradients, and updates its loss scale. MWMS currently does not support allreducing in fp16, since then you must allreduce scaled gradients, but each worker may have a different loss scale. But at least it supports mixed precision for the forwards + backwards pass of the model with loss scaling.

With this proposal, extra work will have to be done even if we don't care about transferring gradients in fp16. We could have each worker have its own loss scale, like in MWMS, but this is more unnatural and difficult since all other variables are on the PS.

Reply:
What if the loss scale is also synchronized, like mean/variance in SyncBatchNormalization? Like you said, it could be aggregated over workers in MWMS or stored on PS.

Reply:
I didn't know SyncBatchNormalization existed until now, but I just read the documentation. I'm not sure how SyncBatchNormalization would work with this proposal either. The [SyncBatchNormalization documentation](https://www.tensorflow.org/api_docs/python/tf/keras/layers/experimental/SyncBatchNormalization) states it synchronizes "the global batch statistics across all devices that are training the model". But you cannot synchronizes with this proposal as it uses asynchronous training. In general, I'm not sure what happens if you call `all_reduce` with a PS Strategy.

I think aggregating the loss scale over workers is tricky since they are asynchronous, but storing it on the PS would work, I think.

Reply:
For common use case in which the batch norm statistics are updated using exponential decay, storing them on PS and updating asynchronously might work without too much harm to the accuracy.

But I agree calling AllReduce in a PS strategy is weird. Ping @anj-s who implemented SyncBN.